File: Recursion/Recursion.cpp
#include<iostream>
#include<vector>
using namespace std;

int coinInfiniteCombinations(vector<int>& coins,int idx, int tar,string ans){
    if(tar==0){
        cout<<ans<<endl;
        return 1;
    }
    int count=0;
    for(int i=idx;i<coins.size();i++){
        if(tar-coins[i]>=0){
            count+=coinInfiniteCombinations(coins,i,tar-coins[i],ans + to_string(coins[i]) + " ");
        }
    }    
    return count;
}

int CoinPermutation(vector<int> &coins, vector<bool> &vis, int tar, string ans)
{
    if (tar == 0)
    {
        cout << ans << endl;
        return 1;
    }

    int count = 0;
    for (int i = 0; i < coins.size(); i++)
    {
        if (tar - coins[i] >= 0 && !vis[i])
        {
            vis[i] = true;
            count += CoinPermutation(coins, vis, tar - coins[i], ans + to_string(coins[i]) + " ");
            vis[i] = false;
        }
    }

    return count;
}
int coinInfinitePermutation(vector<int>& coins, int tar,string ans){
    if(tar==0){
        cout<<ans<<endl;
        return 1;
    }
    int count=0;
    for(int i=0;i<coins.size();i++){
        if(tar-coins[i]>=0){
            count+=coinInfinitePermutation(coins,tar-coins[i], ans + to_string(coins[i]) + " ");
        }
    }    
    return count;
}

int queenCombinations(vector<int> &board,int idx,int tq,int qpsf,string ans){
    if(tq==qpsf){
        cout<<ans<<endl;
        return 1;
    }
    int count=0;
  
    for(int i=idx;i<board.size();i++){
        count+=queenCombinations(board , i+1 , tq,qpsf+1 ,ans+ "q"+to_string(qpsf)+"b"+to_string(i)+"  ");
    }
    return count;
}

int queenPermutations(vector<int> &board,vector<bool> &vis,int tq,int qpsf,string ans){
    if(tq==qpsf){
        cout<<ans<<endl;
        return 1;
    }
    int count=0;
  
    for(int i=0;i<board.size();i++){
        if(!vis[i]){
            vis[i]=true;
            count+=queenPermutations(board , vis , tq,qpsf+1 ,ans+ "q"+to_string(qpsf)+"b"+to_string(i)+"  ");
            vis[i]=false;
    }
    }
    return count;
}
vector<vector<int>> dir_for_nqueen={{0,-1},{-1,-1},{-1,0},{-1,1}};
bool isQueenSafe(vector<vector<int>> &boxes,int r,int c){
    int nr=0,nc=0;
   for(int dir=0;dir<dir_for_nqueen.size();dir++){
       for(int jump=1;jump<=boxes.size()-1,jump++){
           nr=r + jump*dir_for_nqueen[dir][0];
           nc=c + jump*dir_for_nqueen[dir][1];
           if(boxes[nr][nc]==1){
               return false; 
           }
       }
   } 
   return true;
}
int nQueen(vector<vector<int>> &boxes,int idx,int tq, int qpsf, string ans){
if(tq==qpsf){
    cout<<ans<<endl;
    return 1;
}


int count=0;
for(int i=idx;i<boxes.size()*boxes[0].size();i++){
    int r= i/boxes.size();
    int c= i%boxes.size();
    if(isQueenSafe(boxes,r,c)){
        boxes[r][c]=1;
        count+=nQueen(boxes,i+1,tq,qpsf+1,ans +"q"+ to_string(qpsf)+ "r"+ to_string(r) + "c" + to_string(c) +" ");
        boxes[r][c]=0;
    }
}
return count;
}



bool isSafe(int nsr,int nsc,vector<vector<int>>& board){
    if(nsr<0 || nsc<0 || nsr>=board.size() || nsc>=board[0].size() || board[nsr][nsc]==1)
        return false;
    return true;
}
vector<vector<int>> dir={{-1,0},{1,0},{0,1},{0,-1}};
vector<char> directions={'U','D','R','L'};

int floodFill(vector<vector<int>>& board,int sr,int sc,int er,int ec,string ans){
if(sr==er && sc==ec){
    cout<<ans<<endl;
    return 1;
}
int count=0;
board[sr][sc]=1;
for(int d=0;d<dir.size();d++){
    int nsr=sr+dir[d][0];
    int nsc=sc+dir[d][1];
    if(isSafe(nsr,nsc,board)){
        count+=floodFill(board,nsr,nsc,er,ec,ans +directions[d] +" ");
    }
}
board[sr][sc]=0;
return count;
}


// board[sr][sc]=0;
// return count;
// }
// int row_=0;
// int col_=0;
// int diag_=0;
// int antiDiag_=0;

// bool isQueenSafe_bestWala(int col){
// if()
// }

// void mark_unmark_queen(int row,int col){
// row_ ^=(1<<row);
// col_ ^=(1<<col);
// diag_ ^=()
// }

// int nQueen_Best_Using_Bits(int row,int tnq,string ans){
//     if(tnq==0){
//         cout<<ans<<endl;
//         return 1;
//     }
//     int count=0;
//     for(int col=0;col<col_;col++){
//         if(isQueenSafe_bestWala()){
//             mark_unmark_queen(row,col);
//             count+=nQueen_Best_Using_Bits(row+1,tnq-1,ans + "(" + to_string(row) + "," + to_string(col) + ")");
//             mark_unmark_queen(row,col);
//         }
//     }
// return count;
// }
int main(){
    //vector<int> coins={2,3,5,7};
    vector<int> board(5,0);
    vector<bool> vis(5,false);
    vector<vector<int>> boxes();
    //cout<<coinInfinitePermutation(coins,10,"");
    //cout<<queenPermutations(board,vis,3,0,"");
    //cout<<nQueen()
    return 0;
}

File: Recursion/Recursion.java
package Recursion;

import java.util.ArrayList;
class Recursion{

    //==============================Permutations of a string==========================//

    public static void permutation_01(String str,String ans) {
        if(str.length()==0){
            System.out.println(ans);
            return;
        }

        for(int i=0;i<str.length();i++){
            char ch=str.charAt(i);
            String nstr= str.substring(0,i) + str.substring(i+1);
            permutation_01(nstr, ans + ch);
        }
    }


    public static void permutation_02(String str, String ans) {
        if(str.length()==0){
            System.out.println(ans);
            return;
        }
        int[] vis=new int[26];
        for(int i=0;i<str.length();i++){
        String nstr= str.substring(0,i)+str.substring(i+1);
        char ch=str.charAt(i);
        if(vis[ch-'a']!=1){
            vis[ch-'a']=1;
            permutation_02(nstr, ans + ch);
        }

    }
    }
public static ArrayList<String> substr(String str){
if(str.length()==0){
    ArrayList<String> base= new ArrayList<>();
    base.add("");
    return base;
}

String nstr=str.substring(1);
char ch=str.charAt(0);
ArrayList<String> recAns=substr(nstr);
ArrayList<String> myAns= new ArrayList<>(recAns);
for(int i=0;i<recAns.size();i++){
    myAns.add(ch + recAns.get(i));
}
return myAns;
}

public static Boolean isSafe(int nsr,int nsc,int[][] board) {
    if(nsr<0 || nsc<0 || nsr>=board.length || nsc>=board[0].length || board[nsr][nsc] == 1){
        return false;
    }
    return true;
}
public static int[][] dir={{-1,0},{1,0},{0,1},{0,-1}};
public static char[] directions={'U','D','R','L'};
public static int floodFill(int[][] board,int sr,int sc,int er,int ec,String ans) {
    if(sr==er && sc==ec){
        System.out.println(ans);
        return 1;
    }
    int count=0;
    board[sr][sc]=1;
    for(int i=0;i<dir.length;i++){
            int nsr=sr +dir[i][0];
            int nsc=sc +dir[i][1];
            if(isSafe(nsr,nsc,board)){
                count+=floodFill(board, nsr, nsc, er, ec, ans+ directions[i] + " ");
            
        }
    }
    board[sr][sc]=0;
    return count;
}
//========================RAT IN A MAZE================================//
public static void maze() {
    int[][] maze={{1, 0, 0, 0} , {1, 1, 0, 1} , {0, 1, 0, 0} , {1, 1, 1, 1}};
    //System.out.println(maze_01(maze,0,0,3,3,""));
    System.out.println(maze_02(maze,0,0,3,3,""));

}
public static Boolean isSafe_forMaze(int[][] maze,int nsr,int nsc,int er,int ec){
if(nsr<0 || nsc<0 || nsr>=maze.length || nsc>=maze[0].length || maze[nsr][nsc]==0)
    return false;
return true;
}
public static int[][] dir_for_maze={{0,1},{1,0}};
public static String[] directions_for_maze={"H","V"};
public static Boolean maze_01(int[][] maze,int sr,int sc,int er ,int ec,String ans) { //Horizontal & vertical
if(sr==er && sc==ec){
    System.out.println(ans);
    return true;
}
Boolean pathFound=false;
for(int i=0;i<dir_for_maze.length;i++){
int nsr = sr + dir_for_maze[i][0];
int nsc= sc + dir_for_maze[i][1];
if(isSafe_forMaze(maze, nsr, nsc, er, ec)){
   pathFound= maze_01(maze, nsr, nsc, er, ec, ans + directions_for_maze[i] + " ");
}
}
return pathFound;
}



public static int[][] dir_for_maze02={{0,1},{1,0},{1,1}};
public static String[] directions_for_maze02={"H","V","D"};
public static Boolean maze_02(int[][] maze,int sr,int sc,int er ,int ec,String ans) { //Horizontal vertical & diag
    if(sr==er && sc==ec){
        System.out.println(ans);
        return true;
    }
    Boolean pathFound=false;
    for(int i=0;i<dir_for_maze02.length;i++){
    int nsr = sr + dir_for_maze02[i][0];
    int nsc= sc + dir_for_maze02[i][1];
    if(isSafe_forMaze(maze, nsr, nsc, er, ec)){
       pathFound= maze_02(maze, nsr, nsc, er, ec, ans + directions_for_maze02[i] + " ");
    }
    }
    return pathFound;
    }

    public static void nQueenBit_best(int tnq,int rows,String ans) {
        
    }
 public static void solve() {
    //System.out.println(substr("abc"));
    //permutation_02("aaa", "");
    //System.out.println(floodFill(new int[4][4], 0, 0, 3, 3, ""));
    maze();
}

public static void main(String[] args) {
   solve();
}
}